waspc/src/Wasp/Generator/Setup.hs

```diff
@@ -9,13 +9,21 @@ import Wasp.Generator.Common (ProjectRootDir)
 import qualified Wasp.Generator.DbGenerator as DbGenerator
 import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
 import Wasp.Generator.NpmInstall (ensureNpmInstall)
+import Wasp.Message (MessageType (..), SendMessage)
 
-runSetup :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
-runSetup spec dstDir = do
-  -- TODO: Use a monad transformer here in the future for better shortcircuiting behavior.
+runSetup :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> SendMessage -> IO ([GeneratorWarning], [GeneratorError])
```

Comment:
now looking at this pattern, this might in the future be a good candidate for sth like WriterT. Def not know and I like it being this explicit (plus I am not really sure yet how Writer works, and I've heard it should be avoided for some performance reasons actually), but just wanted to share the idea of abstracting it (in the future, once again :D).

Reply:
Yes I think this is correct general direction! While WriterT is typical example for collecting logs, these days it is even more popular to use ReaderT for everything, but I agree with the idea that we could have a monad transformer stack that exposes method `sendMessage` (and it would probably use ReaderT that would contain this same function as dependency injection).

I already discussed this with @faassen previously and we recognized that as an option, but decided to leave it for later.

Just to provide some more context, we already do have monad transformer stack called `Generator`, but that one is pure in the sense that it doesn't use IO. For the parts of generator that are not pure, and from which these kind of messages can be sent, we could create another monad transformer stack, called `GeneratorIO` or smth like that, and it would do what I described at the beginning of this comment.